File: Frontend/src/domain/entities/Jugador.ts
/**
 * Entidad de dominio que representa a un jugador de la partida.
 *
 * @remarks
 * El {@link Jugador.connectionId | connectionId} permite identificar de forma única al
 * jugador local (más fiable que el nombre, que puede repetirse).
 */

import { ID, Color } from '../../core/types';

export class Jugador {
  id: ID;
  nombre: string;
  /** Color asignado por el servidor, o `null` hasta que se asigne. */
  color: Color | null;
  /** ConnectionId de SignalR; `null` si no se conoce. Sirve para identificar al jugador local. */
  connectionId: string | null;

  constructor(id: ID, nombre: string, color: Color | null = null, connectionId: string | null = null) {
    this.id = id;
    this.nombre = nombre;
    this.color = color;
    this.connectionId = connectionId;
  }

  /** Asigna el color (bando) del jugador. */
  asignarColor(color: Color): void {
    this.color = color;
  }

  /**
   * Serializa la entidad a un objeto plano.
   *
   * @remarks
   * No incluye `connectionId` (dato de transporte, no de dominio).
   * @returns Objeto con `id`, `nombre` y `color`.
   */
  toPlain(): object {
    return {
      id: this.id,
      nombre: this.nombre,
      color: this.color,
    };
  }

  /**
   * Construye un {@link Jugador} a partir de un DTO del servidor.
   *
   * @remarks
   * Tolera claves en camelCase y PascalCase (`id`/`Id`/`ID`, `color`/`Color`, …).
   *
   * @param dto - DTO de jugador recibido por SignalR.
   * @returns La entidad de dominio equivalente.
   * @throws Error si el DTO es nulo o le faltan `id`/`nombre`.
   */
  static createFromDTO(dto: any): Jugador {
    if (!dto) {
      throw new Error('DTO de Jugador vacío');
    }

    // Aceptar tanto PascalCase como camelCase
    const id = dto.id ?? dto.Id ?? dto.ID ?? null;
    const nombre = dto.nombre ?? dto.Nombre ?? null;
    const color = dto.color ?? dto.Color ?? null;
    const connectionId = dto.connectionId ?? dto.ConnectionId ?? null;

    if (!id || !nombre) {
      throw new Error('DTO incompleto para crear Jugador');
    }

    return new Jugador(String(id), String(nombre), color ?? null, connectionId ?? null);
  }
}

File: Backend/Domain/Entities/Jugador.cs
using ServidorAjedrez.Domain.Enums;

namespace ServidorAjedrez.Domain.Entities
{
    /// <summary>
    /// Jugador participante en una sala o partida.
    /// </summary>
    /// <remarks>
    /// Maneja dos identificadores con propósitos distintos: <see cref="Id"/> es la identidad
    /// estable del jugador dentro del dominio, mientras que <see cref="ConnectionId"/> es el
    /// identificador de transporte de SignalR (puede cambiar al reconectar). Buena parte del
    /// flujo del servidor traduce el connectionId entrante al <see cref="Id"/> de dominio. El
    /// color permanece sin asignar hasta que la partida reparte los bandos.
    /// </remarks>
    public class Jugador
    {
        private string _id;
        private string _nombre;
        private string _connectionId;
        private Color? _color;

        /// <summary>Identidad estable del jugador en el dominio (GUID generado al crearlo).</summary>
        public string Id => _id;

        /// <summary>Nombre visible elegido por el jugador.</summary>
        public string Nombre => _nombre;

        /// <summary>
        /// Identificador de la conexión SignalR del jugador, usado para enrutar mensajes y para
        /// resolver su <see cref="Id"/>. Puede cambiar si el cliente se reconecta.
        /// </summary>
        public string ConnectionId => _connectionId;

        /// <summary>Bando asignado al jugador, o <c>null</c> mientras la partida no lo haya repartido.</summary>
        public Color? Color => _color;

        /// <summary>
        /// Crea un jugador con identidad nueva y sin color asignado.
        /// </summary>
        /// <param name="nombre">Nombre visible; no puede ser nulo ni estar en blanco.</param>
        /// <param name="connectionId">Identificador de la conexión SignalR; no puede ser nulo ni estar en blanco.</param>
        /// <exception cref="ArgumentException">
        /// <paramref name="nombre"/> o <paramref name="connectionId"/> son nulos o están en blanco.
        /// </exception>
        public Jugador(string nombre, string connectionId)
        {
            if (string.IsNullOrWhiteSpace(nombre))
                throw new ArgumentException("El nombre del jugador no puede estar vacío.");
            if (string.IsNullOrWhiteSpace(connectionId))
                throw new ArgumentException("El connectionId no puede estar vacío.");

            _id = Guid.NewGuid().ToString();
            _nombre = nombre;
            _connectionId = connectionId;
            _color = null;
        }

        /// <summary>Asigna el bando con el que el jugador disputa la partida.</summary>
        /// <param name="color">Bando que se adjudica al jugador.</param>
        public void AsignarColor(Color color)
        {
            _color = color;
        }

        /// <summary>
        /// Crea un jugador. Equivale al constructor; se ofrece como método de fábrica por
        /// consistencia con el resto de entidades del dominio.
        /// </summary>
        /// <param name="nombre">Nombre visible; no puede ser nulo ni estar en blanco.</param>
        /// <param name="connectionId">Identificador de la conexión SignalR; no puede ser nulo ni estar en blanco.</param>
        /// <returns>La nueva instancia de <see cref="Jugador"/>.</returns>
        /// <exception cref="ArgumentException">
        /// <paramref name="nombre"/> o <paramref name="connectionId"/> son nulos o están en blanco.
        /// </exception>
        public static Jugador Create(string nombre, string connectionId)
        {
            return new Jugador(nombre, connectionId);
        }
    }
}
